### PR TITLE
[ci, vllm] test: add CI for process_weights_after_loading correctness

### PR DESCRIPTION
**Pull request by Kamleecoder**
### What does this PR do?

This PR adds a CI test to verify the correctness of `vllm.model_executor.model_runner.model_loader.utils.process_weights_after_loading`.

Some DeepSeek-like models (e.g., Moonlight-16B-A3B) use **Multi-head Latent Attention (MLA)** with `q_lora_rank` enabled. When `process_weights_after_loading` is not properly implemented, these models produce **garbled output** during inference. This CI test guards against such regressions.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True, torch_dtype="auto")
+    try:
+        yield from model.state_dict().items()
+    finally:
+        del model
+        gc.collect()
+
+
+def _update_weights(server, model_path: str):
+    zmq_handle = f"ipc:///tmp/rl-colocate-zmq-{get_device_uuid(get_device_id())}.sock"
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py]: 已按 gemini 建议修改，提交在 commit ，/gemini review 触发请再看一次

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+from verl.utils.device import get_device_id, is_support_ipc
+from verl.utils.tokenizer import normalize_token_ids
+from verl.workers.rollout.replica import RolloutMode, TokenOutput
+from verl.workers.rollout.vllm_rollout.bucketed_weight_transfer import BucketedWeightSender
+from verl.workers.rollout.vllm_rollout.utils import get_device_uuid
+from verl.workers.rollout.vllm_rollout.vllm_async_server import vLLMHttpServer
+
+MODEL_PATH = "/data02/Moonlight-16B-A3B-Instruct"
+MODEL_PATH_QWEN3_14B = "/data02/Qwen3/Qwen3-14B"
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+    if force_dummy:
+        ray.get(server.__ray_call__.remote(lambda self: setattr(self.config, "load_format", "dummy")))
+
+    ray.get(server.launch_server.remote())
+    return server
+
+
+def _iter_weights(model_path: str):
+    model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True, torch_dtype="auto")
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:128]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        .remote(
+            config=rollout_cfg,
+            model_config=model_cfg,
+            rollout_mode=RolloutMode.STANDALONE,
+            workers=[],
+            replica_rank=0,
+            node_rank=0,
+            gpus_per_node=1,
+            nnodes=1,
+            cuda_visible_devices="6",
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+    if force_dummy:
+        ray.get(server.__ray_call__.remote(lambda self: setattr(self.config, "load_format", "dummy")))
+
+    ray.get(server.launch_server.remote())
+    return server
+
+
+def _iter_weights(model_path: str):
+    model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True, torch_dtype="auto")
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:128]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        yield from model.state_dict().items()
+    finally:
+        del model
+        gc.collect()
+
+
+def _update_weights(server, model_path: str):
+    # Use the same zmq handle format as vLLMColocateWorkerExtension._get_zmq_handle()
+    replica_rank = os.environ.get("VERL_REPLICA_RANK", "0")
+    zmq_handle = f"ipc:///tmp/rl-colocate-zmq-replica-{replica_rank}-rank-0.sock"
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:141]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            replica_rank=0,
+            node_rank=0,
+            gpus_per_node=1,
+            nnodes=1,
+            cuda_visible_devices="6",
+        )
+    )
+
+    if force_dummy:
+        ray.get(server.__ray_call__.remote(lambda self: setattr(self.config, "load_format", "dummy")))
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:122]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        .remote(
+            config=rollout_cfg,
+            model_config=model_cfg,
+            rollout_mode=RolloutMode.STANDALONE,
+            workers=[],
+            replica_rank=0,
+            node_rank=0,
+            gpus_per_node=1,
+            nnodes=1,
+            cuda_visible_devices="6",
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+    if force_dummy:
+        ray.get(server.__ray_call__.remote(lambda self: setattr(self.config, "load_format", "dummy")))
+
+    ray.get(server.launch_server.remote())
+    return server
+
+
+def _iter_weights(model_path: str):
+    model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True, torch_dtype="auto")
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:128]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            replica_rank=0,
+            node_rank=0,
+            gpus_per_node=1,
+            nnodes=1,
+            cuda_visible_devices="0",
+        )
+    )
+
+    if force_dummy:
+        ray.get(server.__ray_call__.remote(lambda self: setattr(self.config, "load_format", "dummy")))
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:122]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    runtime_env = {
+        "TOKENIZERS_PARALLELISM": "true",
+        "VERL_LOGGING_LEVEL": "INFO",
+        "VLLM_LOGGING_LEVEL": "INFO",
+        "HCCL_CONNECT_TIMEOUT": os.environ.get("HCCL_CONNECT_TIMEOUT", "1500"),
+        "HCCL_HOST_SOCKET_PORT_RANGE": os.environ.get("HCCL_HOST_SOCKET_PORT_RANGE", "60000-60050"),
+        "HCCL_NPU_SOCKET_PORT_RANGE": os.environ.get("HCCL_NPU_SOCKET_PORT_RANGE", "61000-61050"),
+    }
+    if not ray.is_initialized():
+        ray.init(runtime_env={"env_vars": runtime_env}, ignore_reinit_error=True)
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py]: /gemini review

File: tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        yield from model.state_dict().items()
+    finally:
+        del model
+        gc.collect()
+
+
+def _update_weights(server, model_path: str):
+    # Use the same zmq handle format as vLLMColocateWorkerExtension._get_zmq_handle()
+    replica_rank = os.environ.get("VERL_REPLICA_RANK", "0")
+    zmq_handle = f"ipc:///tmp/rl-colocate-zmq-replica-{replica_rank}-rank-0.sock"
```
[the remark this answers is not in the document]
Comment (Kamleecoder) [reply tests/workers/rollout/rollout_vllm/test_vllm_process_weights_after_loading.py:141]: /gemini review

**Comment by tardis-key** [comment]
You need to ensure that your test.py is included in a workflow, so that the test script can be executed in CI.(I'm not sure if it will be included in any pytest workflow like `pytest tests/*`)

**Comment by tardis-key** [comment]
How long does it take to run these unit tests? UTs need to be as efficient as possible.

**Comment by Kamleecoder** [comment]
> You need to ensure that your test.py is included in a workflow, so that the test script can be executed in CI.(I'm not sure if it will be included in any pytest workflow like `pytest tests/*`)
